content/draft/general/password_strength.md

```diff
@@ -0,0 +1,91 @@
+
+* [designs](https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba%E2%8E%9CDesktop?node-id=11%3A4921)
+
+## Use Cases
+
+### Account creation
+
+1. User types the choosen password in the password field
+2. System displays feedback on the strength of the password
+2.1 System checks if the password fits criteria (see Functional Requirements#password criteria) and displays feedback (e.g "Password is too short")
+2.2 System categorizes password strength (see Functional Requirements#password categorization) and displays matching category (e.g password Is "Good")
+2.2 System checks if password is a common password (see Functional Requirements#password validation) and displays feedback (e.g "this pasword has been pwned and shouldn't be used")
+3. User types password again in the confirm password field
+4. System checks if password matches and displays "Passwords don't match" if they don't
+5. System checks if password matches and enables the "Create" button if they do
+6. User clicks "Create" button
+
+### Change password
+
+1. User types the current password in the current password field
+2. User types the choosen password in the new password field
+3. System displays feedback on the strength of the password
+4.1 System checks if the password fits criteria (see Functional Requirements#password criteria) and displays feedback (e.g "Password is too short")
+4.2 System categorizes password strength (see Functional Requirements#password categorization) and displays matching category (e.g password Is "Good")
+4.2 System checks if password is a common password (see Functional Requirements#password validation) and displays feedback (e.g "this pasword has been pwned and shouldn't be used")
+5. User types password again in the confirm password field
+6. System checks if password matches and displays "Passwords don't match" if they don't
+7. System checks if password matches and enables the "Create" button if they do
+8. User clicks "Create" button
+7. System checks if current password is valid and displays "Current password is invalid" if it isn't
+
+## Functional Requirements
+
+### Password criteria
+- A password MUST BE at least 6 characters long.
+- A password MUST BE at most 64 characters long.
+- A password MUST NOT contain spaces.
+- A password MAY CONTAIN any printable ascii character.
+
+### Password validation
+
+- A password MUST be checked against a list of common passwords
+    - If a password matches a common password, it MUST BE rejected
+    - If a password matches a common password, it MUST BE display a warning "this pasword has been pwned and shouldn't be used"
+    - If the application is online, the password MAY BE checked against an online or a local database.
+    - If the application is offline, the password MAY BE checked against a local database.
+
+### Password categorization
+- A Password MUST BE categorized as one of the following:
+  - Very Weak
+  - Weak
+  - So-So
+  - Good
+  - Great
+- **TODO**: criteria for each category
```

Comment:
@corpetty @0kok0 we need to define this sort of categorization by some sort of formula and/or criteria

Reply:
Corey and co. prolly have much better ideas here, but one would be to go for:

```
NUM_CRITERIA = amount of criteria we have for validation
PASSWORD_CRITERIA_MATCHES = amount of matching criteria in given password

MATCHES_IN_PERCENTAGE = (PASSWORD_CRITERIA_MATCHES/NUM_CRITERIA)*100

match MATCHES_IN_PERCENTAGE:
  100 -> Great
  >= 80 -> Good
  >= 60 -> So-So
  >= 40 -> Weak
  < 40 -> Very Weak
```

Reply:
> BTW, is there any restriction for using 6 characters minimum, as well as, not allowing space character? From [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls) the minimum length should be 8 and space char should be allowed to have a better strength password.

Yes, the consensus is recent years has been that [we should focus on length of the passwords, allow spaces (and if possible unicode) & avoid complex composition rules](https://nakedsecurity.sophos.com/2016/08/18/nists-new-password-rules-what-you-need-to-know/)

Reply:
Also, I'd like to add an excellent article about it:

https://www.ndss-symposium.org/wp-content/uploads/2017/09/06_3_1.pdf

Reply:
@noeliaSD @OxFred @John-44 

As far as I know, the best algorithm for the Password Strength Estimation is [Dropbox's zxcvbn](https://github.com/dropbox/zxcvbn).

Bitwarden is also [using zxcvbn](https://github.com/bitwarden/jslib/blob/c7ac645eb7b0cfa9c972dd382d63fab5893e1f82/src/services/passwordGeneration.service.ts#L383). So we should consider using it.

Reply:
As does nimbus: https://github.com/status-im/nim-zxcvbn

Reply:
> @noeliaSD @OxFred @John-44
> 
> As far as I know, the best algorithm for the Password Strength Estimation is [Dropbox's zxcvbn](https://github.com/dropbox/zxcvbn).
> 
> Bitwarden is also [using zxcvbn](https://github.com/bitwarden/jslib/blob/c7ac645eb7b0cfa9c972dd382d63fab5893e1f82/src/services/passwordGeneration.service.ts#L383). So we should consider using it.

Yes sure ! that's what was initially suggested on [the security-internal issue](https://github.com/status-im/security-internal/issues/56#issuecomment-984921340)